From 0161adf28590d95eaf234acc97156bc7dd989c8c Mon Sep 17 00:00:00 2001
From: Anton Gerasimov <anton@advancedtelematic.com>
Date: Wed, 28 Dec 2016 16:53:35 +0100
Subject: [PATCH] Load environment from the boot partition

---
 include/configs/minnowmax.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/configs/minnowmax.h b/include/configs/minnowmax.h
index 3aa22d2..02f6980 100644
--- a/include/configs/minnowmax.h
+++ b/include/configs/minnowmax.h
@@ -41,4 +41,7 @@
 #define CONFIG_ENV_SECT_SIZE		0x1000
 #define CONFIG_ENV_OFFSET		0x006ef000
 
+#undef CONFIG_BOOTCOMMAND
+#define CONFIG_BOOTCOMMAND "fatload mmc 1:1 0x01000000 uEnv.txt; env import -t 0x01000000 ${filesize}; run bootcmd"
+
 #endif	/* __CONFIG_H */
-- 
2.9.3

